### PR TITLE
[`split_special_tokens`] Add support for `split_special_tokens` argument to encode

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Argument name is totally debatable. Will also require a pull request in `tokenizers`. 
The goal is to be able to simply activate and de-activate the special token splitting. Feature was asked in #22490, and is required for some production type cases, where users pass inputs and we don't want them to be able to hack them

### CODE REVIEW

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -513,8 +515,14 @@ def tokenize(self, text: TextInput, **kwargs) -> List[str]:
...
 
-        no_split_token = set(self.unique_no_split_tokens)
-        tokens = self.tokens_trie.split(text)
+        # split_special_tokens: empty `no_split_token`
+        if split_special_tokens:
+            no_split_token = []
+            tokens = [text]
+        else:
+            no_split_token = set(self.unique_no_split_tokens)
+            tokens = self.tokens_trie.split(text)
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils.py:524]: Luckily enough, this is the only change for slow tokenizers.

Comment (ArthurZucker) [reply src/transformers/tokenization_utils.py:524]: Might be a cleaner/better solution, will check

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -2157,7 +2165,7 @@ def save_pretrained(
 
         # TODO: Ensure the modified attributes (those are also in the __init__ kwargs) will give identical tokenizers
         # target_keys = self.init_kwargs.keys()
-        target_keys = ["model_max_length", "clean_up_tokenization_spaces"]
+        target_keys = ["model_max_length", "clean_up_tokenization_spaces", "split_special_tokens"]
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:2168]: Make sure the argument is added to the keys to save/load

File: src/transformers/tokenization_utils_base.py [hunk 3]
```diff
@@ -1492,6 +1492,11 @@ def all_special_ids(self) -> List[int]:
         clean_up_tokenization_spaces (`bool`, *optional*, defaults to `True`):
             Whether or not the model should cleanup the spaces that were added when splitting the input text during the
             tokenization process.
+        split_special_tokens (`bool`, *optional*, defaults to `False`):
+            Whether or not the special tokens should be split during the tokenization process. The default behaviour is
+            to not split special tokens. This means that if `<s>` is the `bos_token`, then `tokenizer.tokenize("<s>") =
+            ['<s>`]. Otherwise, if `split_special_tokens=True`, then `tokenizer.tokenize("<s>")` will be give `['<',
+            's', '>']`.
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/tokenization_utils_base.py:1499]: Yes, that'd be best!

**Review by sgugger — APPROVED** [review]
Thanks again!

**Pull request merged**